src/relay/backend/contrib/dnnl/codegen.cc

```diff
@@ -128,42 +128,50 @@ std::vector<std::string> Add(const CallNode* call) {
 
 // TODO(@zhiics, @comaniac): This is a basic implementation. We should implement
 // all utilities and make a base class for users to implement.
-class CodegenDNNL : public ExprVisitor, public CodegenCBase {
+class CodegenDNNL : public relay::ExprFunctor<std::vector<Output>(const Expr&)>,
+                    public CodegenCBase {
  public:
   explicit CodegenDNNL(const std::string& id) { this->ext_func_id_ = id; }
 
-  void VisitExpr_(const VarNode* node) final {
+  std::vector<Output> VisitExpr(const Expr& expr) final {
+    if (visited_.count(expr)) return visited_.at(expr);
```

Comment:
I wonder why `ExprFunctor` has memoization built in? It would be nice if derived classes don't have to take care of this.

Reply:
Here is dup-ed logic in another derived class
https://github.com/apache/incubator-tvm/blob/00014e20c3cc077727d467a67d3498260627e4e0/src/relay/backend/compile_engine.cc#L182-L191

Reply:
ExprFunctor actually only has the pure virtual functions. It's developer's responsibility to implement them. Although we could have a map like `map<Expr, R>`, it probably makes sense to let users define your own memoization.

Reply:
hmm but it does have one dispatching function 
https://github.com/apache/incubator-tvm/blob/master/include/tvm/relay/expr_functor.h#L93

Why can't we have memoization there? If a derived class wants a custom memoization it can overide this method.

Reply:
yeah, we could do that, and we can do that in a separate PR if you feel necessary.

Reply:
Well it is not necessary, but it would be nice if we could add memo there. We don't have to add this overide in `CodegenDNN` and we can remove many duped logic in other derived classes.

But since that would be a big change, I agree that we should proceed with this impl for now.